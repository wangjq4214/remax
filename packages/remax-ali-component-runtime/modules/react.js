import { r as react } from './index-c3dc9930.js';

var Children = react.Children;
var Component = react.Component;
var Fragment = react.Fragment;
var Profiler = react.Profiler;
var PureComponent = react.PureComponent;
var StrictMode = react.StrictMode;
var Suspense = react.Suspense;
var cloneElement = react.cloneElement;
var createContext = react.createContext;
var createElement = react.createElement;
var createFactory = react.createFactory;
var createRef = react.createRef;
var forwardRef = react.forwardRef;
var isValidElement = react.isValidElement;
var lazy = react.lazy;
var memo = react.memo;
var unstable_SuspenseList = react.unstable_SuspenseList;
var unstable_withSuspenseConfig = react.unstable_withSuspenseConfig;
var useCallback = react.useCallback;
var useContext = react.useContext;
var useDebugValue = react.useDebugValue;
var useEffect = react.useEffect;
var useImperativeHandle = react.useImperativeHandle;
var useLayoutEffect = react.useLayoutEffect;
var useMemo = react.useMemo;
var useReducer = react.useReducer;
var useRef = react.useRef;
var useState = react.useState;
var version = react.version;
export {
  Children,
  Component,
  Fragment,
  Profiler,
  PureComponent,
  StrictMode,
  Suspense,
  cloneElement,
  createContext,
  createElement,
  createFactory,
  createRef,
  forwardRef,
  isValidElement,
  lazy,
  memo,
  unstable_SuspenseList,
  unstable_withSuspenseConfig,
  useCallback,
  useContext,
  useDebugValue,
  useEffect,
  useImperativeHandle,
  useLayoutEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
  version,
};
