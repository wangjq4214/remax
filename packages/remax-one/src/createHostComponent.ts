import * as React from 'react';
import {
  TapEvent,
  TouchEvent,
  EventTarget,
  EventCurrentTarget,
  ImageLoadEvent,
  ImageErrorEvent,
  InputEvent,
  FormEvent,
} from './types';
import { formatDisplayName } from '@remax/shared';

export function createTarget(target: any, detail: any): EventTarget {
  return {
    id: target.id,
    offsetLeft: target.offsetLeft,
    offsetTop: target.offsetTop,
    dataset: target.targetDataset || target.dataset,
    value: detail?.value,
  };
}

export function createCurrentTarget(currentTarget: any): EventCurrentTarget {
  return {
    id: currentTarget.id,
    offsetLeft: currentTarget.offsetLeft,
    offsetTop: currentTarget.offsetTop,
    dataset: currentTarget.dataset,
  };
}

export const createTapEvent = (originalEvent: any): TapEvent => ({
  type: originalEvent.type,
  stopPropagation: originalEvent.stopPropagation,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export const createTouchEvent = (originalEvent: any): TouchEvent => ({
  type: originalEvent.type,
  stopPropagation: originalEvent.stopPropagation,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  touches: originalEvent.touches,
  changedTouches: originalEvent.touches,
  originalEvent,
});

export const createImageEvent = (originalEvent: any): ImageLoadEvent | ImageErrorEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export function createCallback(fn: Function | undefined, eventCreator: Function) {
  if (typeof fn !== 'function') {
    return undefined;
  }

  return (originalEvent: any) => fn(eventCreator(originalEvent));
}

export const createInputEvent = (originalEvent: any): InputEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

export const createFormEvent = (originalEvent: any): FormEvent => ({
  type: originalEvent.type,
  target: createTarget(originalEvent.target, originalEvent.detail),
  currentTarget: createCurrentTarget(originalEvent.currentTarget),
  originalEvent,
});

function assignDefaultProps(inputProps: any, defaultProps: any) {
  if (defaultProps) {
    Object.keys(defaultProps).forEach(key => {
      inputProps[key] = inputProps[key] ?? defaultProps[key];
    });
  }
}

export function aliasProps(props: any, alias: { [key: string]: string }) {
  if (!alias) {
    return props;
  }

  const nextProps: any = {};

  for (const key in props) {
    nextProps[alias[key] ?? key] = props[key];
  }

  return nextProps;
}

export default function createHostComponent<P = any>(
  name: string,
  alias: {
    [key: string]: string;
  } | null,
  defaults?: {
    [key: string]: any;
  }
) {
  const Component: React.ForwardRefRenderFunction<any, P> = (props: any, ref: React.Ref<any>) => {
    const inputProps = {
      ...props,
    };

    // 默认属性根据平台在这里设置
    if (defaults) {
      assignDefaultProps(inputProps, defaults);
    }

    if (props.onLongTap) {
      inputProps.onLongTap = createCallback(inputProps.onLongTap, createTapEvent);
    }
    if (inputProps.onTap) {
      inputProps.onTap = createCallback(inputProps.onTap, createTapEvent);
    }

    if (inputProps.onTouchStart) {
      inputProps.onTouchStart = createCallback(inputProps.onTouchStart, createTouchEvent);
    }

    if (inputProps.onTouchMove) {
      inputProps.onTouchMove = createCallback(inputProps.onTouchMove, createTouchEvent);
    }

    if (inputProps.onTouchEnd) {
      inputProps.onTouchEnd = createCallback(inputProps.onTouchEnd, createTouchEvent);
    }

    if (inputProps.onTouchCancel) {
      inputProps.onTouchCancel = createCallback(inputProps.onTouchCancel, createTouchEvent);
    }

    if (inputProps.onChange) {
      inputProps.onChange = createCallback(inputProps.onChange, createInputEvent);
    }

    if (inputProps.onInput) {
      inputProps.onInput = createCallback(inputProps.onInput, createInputEvent);
    }

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createCallback(inputProps.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createCallback(inputProps.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createCallback(inputProps.onBlur, createInputEvent);
    }

    if (inputProps.onSubmit) {
      inputProps.onSubmit = createCallback(inputProps.onSubmit, createFormEvent);
    }

    if (inputProps.onReset) {
      inputProps.onReset = createCallback(inputProps.onReset, createFormEvent);
    }

    if (name === 'image') {
      if (inputProps.onLoad) {
        inputProps.onLoad = createCallback(props.onLoad, createImageEvent);
      }
      if (inputProps.onError) {
        inputProps.onError = createCallback(props.onError, createImageEvent);
      }
    }

    let nextProps = inputProps;

    if (alias) {
      nextProps = aliasProps(inputProps, alias);
    }
    nextProps.ref = ref;

    return React.createElement(name, nextProps);
  };

  if (process.env.NODE_ENV === 'development') {
    Component.displayName = formatDisplayName(name);
  }

  return React.forwardRef<any, React.PropsWithChildren<P>>(Component);
}
