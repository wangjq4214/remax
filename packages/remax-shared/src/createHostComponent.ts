import * as React from 'react';
import { formatDisplayName } from './formatDisplayName';

export function createHostComponent<P = any>(name: string, component?: React.ComponentType<P>) {
  if (component) {
    return component;
  }

  const Component: React.ForwardRefRenderFunction<any, P> = (props, ref: React.Ref<any>) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };

  if (process.env.NODE_ENV === 'development') {
    Component.displayName = formatDisplayName(name);
  }

  return React.forwardRef<any, React.PropsWithChildren<P>>(Component);
}
