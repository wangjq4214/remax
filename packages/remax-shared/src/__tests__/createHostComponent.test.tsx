import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import { createHostComponent } from '../createHostComponent';

describe('createHostComponent', () => {
  it('create correctly', () => {
    const RemaxCustomComponent = createHostComponent<{ foo: string }>('custom-component');
    const testRenderer = TestRenderer.create(<RemaxCustomComponent foo="bar" />);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('create with custom render', () => {
    const RemaxCustomComponent = createHostComponent<{ foo: string }>('custom-component', props =>
      React.createElement('input', props)
    );
    const testRenderer = TestRenderer.create(<RemaxCustomComponent foo="bar" />);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
