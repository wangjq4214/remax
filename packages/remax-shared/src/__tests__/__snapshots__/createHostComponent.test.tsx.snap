// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHostComponent create correctly 1`] = `
<custom-component
  foo="bar"
/>
`;

exports[`createHostComponent create with custom render 1`] = `
<input
  foo="bar"
/>
`;
